bundles/org.openhab.binding.automower/src/main/java/org/openhab/binding/automower/internal/rest/api/authentication/AuthenticationApi.java

```diff
@@ -0,0 +1,120 @@
+/**
+ * Copyright (c) 2010-2020 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.automower.internal.rest.api.authentication;
+
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeoutException;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jetty.client.HttpClient;
+import org.eclipse.jetty.client.api.ContentResponse;
+import org.eclipse.jetty.client.api.Request;
+import org.eclipse.jetty.client.util.FormContentProvider;
+import org.eclipse.jetty.http.HttpMethod;
+import org.eclipse.jetty.http.HttpStatus;
+import org.eclipse.jetty.util.Fields;
+import org.openhab.binding.automower.internal.rest.api.HusqvarnaApi;
+import org.openhab.binding.automower.internal.rest.api.authentication.dto.PostOAuth2Response;
+import org.openhab.binding.automower.internal.rest.exceptions.AutomowerCommunicationException;
+import org.openhab.binding.automower.internal.rest.exceptions.UnauthorizedException;
+
+import com.google.gson.JsonSyntaxException;
+
+/**
+ * Gives access to the Husqvarna authentication api
+ *
+ * @author Markus Pfleger - Initial contribution
+ */
+@NonNullByDefault
+public class AuthenticationApi extends HusqvarnaApi {
```

Comment:
This looks like a OAuth2 authentication process. Although I miss the callback part. Where normally the service authenticated calls back on a specific url. But that doesn't seem to be done here.

In openHAB we have a service for handling OAuth2 authentication. You can get the service via `OAuthFactory`. See for example the Spotify binding for how that works. It also persists the refresh key and access key. This means you don't have to implement your own. Can you have a look at it. We would prefer if you use it if this is a OAuth2 compatible process.

Reply:
Yes, it is OAuth2. For sure I will use the existing functionality. Thanks for the hint.

Reply:
Hm, had a look at the Husqvarna documentation: https://developer.husqvarnagroup.cloud/apis/Authentication+API#/readme

I did not work with oauth before. What I implemented was the "Password Grant" which allows a login by username and password. 

Looking at the OAuthFactory I guess this uses the "Authorization Code Grant" workflow. I am not sure how this redirect part works exactly, this is why I implemented the password grant.

Looking at the OAuthFactory 2 urls are needed. tokenUrl and authorizationUrl. I have one where I get the token from, but not sure about the second...

Reply:
In this case you can pass `null` for the `authorizationUrl` and call `getAccessTokenByResourceOwnerPasswordCredentials` on the `OAuthClientService` to get an `AccessTokenResponse` with the refresh/accesstoken. Then with `getAccessTokenResponse()` you get the token when you make a call. If the token is expired this method will first get a new token with the refreshtoken. You can also build in a fallback in case you make a call and the token turns out to be expired. You can call `refreshToken()` to force a refresh.

Reply:
Thx, will try that

Reply:
Thanks for the hint. Works very well. Was able to remove quite some code...